src/coreclr/jit/lower.cpp

```diff
@@ -208,6 +214,95 @@ bool Lowering::IsSafeToContainMem(GenTree* grandparentNode, GenTree* parentNode,
+//    childNode  - some node that is an input to `parentNode`
+//
+// Return value:
+//    true if it is safe to make childNode a contained memory operand.
+//
+bool Lowering::IsSafeToContainMem(GenTree* parentNode, GenTree* childNode) const
+{
+    return IsInvariantInRange(childNode, parentNode);
+}
+
+//------------------------------------------------------------------------
+// IsSafeToContainMem: Checks for conflicts between childNode and grandParentNode
+// and returns 'true' iff memory operand childNode can be contained in grandParentNode.
+//
+// Arguments:
+//    grandParentNode - any non-leaf node
+//    parentNode - parent of `childNode` and an input to `grandParentNode`
+//    childNode  - some node that is an input to `parentNode`
+//
+// Return value:
+//    true if it is safe to make childNode a contained memory operand.
+//
+bool Lowering::IsSafeToContainMem(GenTree* grandparentNode, GenTree* parentNode, GenTree* childNode) const
+{
+    return IsInvariantInRange(childNode, grandparentNode, parentNode);
+}
+
+//------------------------------------------------------------------------
+// IsSafeToMarkRegOptional: Check whether it is safe to mark 'childNode' as
+// reg-optional in 'parentNode'.
+//
+// Arguments:
+//    parentNode - parent of 'childNode'
+//    childNode  - some node that is an input to `parentNode`
+//
+// Return value:
+//    True if it is safe to mark childNode as reg-optional; otherwise false.
+//
+// Remarks:
+//    Unlike containment, reg-optionality can only rarely introduce new
+//    conflicts, because reg-optionality mostly does not cause the child node
+//    to be evaluated at a new point in time:
+//
+//    1. For LIR edges (i.e. anything that isn't GT_LCL_VAR) reg-optionality
+//       indicates that if the edge was spilled to a temp at its def, the parent
+//       node can use it directly from its spill location without reloading it
+//       into a register first. This is always safe as as spill temps cannot
+//       interfere.
+//
+//       For example, an indirection can be marked reg-optional even if there
+//       is interference between it and its parent; the indirection will still
+//       be evaluated at its original position, but if the value is spilled to
+//       stack, then reg-optionality can allow using the value from the spill
+//       location directly. Similarly, GT_LCL_FLD nodes are never register
+//       candidates and can be handled the same way.
+//
+//    2. For GT_LCL_VAR reg-optionality indicates that the node can use the
+//       local directly from its home location. IR invariants guarantee that the
+//       local is not defined between its LIR location and the parent node (see
+//       CheckLclVarSemanticsHelper). That means the only case where it could
+//       interfere is due to it being address exposed. So this is the only unsafe
+//       case.
+//
+bool Lowering::IsSafeToMarkRegOptional(GenTree* parentNode, GenTree* childNode) const
```

Comment:
Is any of this optimization that should also apply to `IsSafeToContainMem`?

That is, iirc there is a "pessimization" for the `InterferesWith` check today with regards to `locals` in that any write is considered interfering, even if its to another local and therefore cannot impact the local being read.

However, based on https://github.com/dotnet/runtime/blob/main/docs/design/specs/Memory-model.md I believe we are allowed to reorder such a non-volatile read of a local assuming there are no volatile writes.

Reply:
Maybe what I'm remembering is just the fact that we couldn't mark it reg-optional and the fix is exactly what is being handled here

Reply:
@SingleAccretion managed to find the discussion we had it some time back on Discord

The general scenario I was thinking about is: https://github.com/dotnet/runtime/pull/76273#issuecomment-1261260839

> Many diffs are cases like the following:
> ```diff
> - add     x0, x0, w26, UXTW
> - ldrb    w0, [x0]
> + ldrb    w0, [x0, w26, UXTW #2]
> ```
> 
> These seem to mostly be due to `ADDEX` not being properly handled in many places while `ADD` always is.
> 
> ---------
> 
> There are a few small regressions such as:
> ```diff
> + lsl     w0, w0, #8
>   ldr     x1, [fp, #0xA8]	// [V192 tmp170]
>   ldrb    w1, [x1, #0x01]
> - add     w1, w1, w0,  LSL #8
> + add     w1, w0, w1
> ```
> 
> Notably we're checking if `childNode->gtGetOp1()` (the value to be shifted) is contained, the original doesn't (but it shouldn't be anyways since we need a register for it).
> 
> We're also checking `IsSafeToContainMem(parentNode, childNode)` where-as the original wasn't. I expect this is the "actual" reason for the "larger diff" since we always do "strict" checking and we don't allow reordering across anything with a "side effect". That being said, there is no actual interdependence between these two and so it is technically "safe" anyways.